### PR TITLE
Add key vault and storage account

### PR DESCRIPTION
**Pull request by jade-codes**
## Description
Added references to AVM modules for storage account and key vault

### CODE REVIEW

**Comment by vyta** [comment]
Nice work @Jade-Codes! A general question, why are the private dns zones being created for storage account by default, but not for the aml workspace?

**Comment by jade-codes** [comment]
> Nice work @Jade-Codes! A general question, why are the private dns zones being created for storage account by default, but not for the aml workspace?

Good shout, I was initially following the other modules for doing this, but I definitely could move that into the module too

File: locals.keyvault.tf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+locals {
+  key_vault_name = replace("kv${var.name}", "/[^a-zA-Z0-9-]/", "")
+  key_vault_endpoints = var.is_private ? toset(["vaultcore"]) : toset([])
```
Comment (Nepomuceno) [inline locals.keyvault.tf]: Love how this work together with private endpoints!

File: main.storage.tf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+module "avm_res_storage_storageaccount" {
+  source = "Azure/avm-res-storage-storageaccount/azurerm"
+  enable_telemetry              = var.enable_telemetry
+  // regex to remove invalid characters from the name
+  name                          = local.storage_account_name
+  resource_group_name           = var.resource_group_name
+  location                      = var.location
+  account_replication_type      = "GRS"
+  account_tier                  = "Standard"
+  account_kind                  = "StorageV2"
```
Comment (Nepomuceno) [inline main.storage.tf]: DO you want those to be variables that cna be controled or do you think that if someone wnats to control this they should pass in their own storage account ?

File: main.tf [hunk 3]
```diff
@@ -10,8 +10,8 @@ resource "azurerm_machine_learning_workspace" "this" {
   location                = var.location
   resource_group_name     = var.resource_group_name
   application_insights_id = azurerm_application_insights.this.id
-  key_vault_id            = var.key_vault_id
-  storage_account_id      = var.storage_account_id
+  key_vault_id            = var.key_vault == null ? module.avm_res_keyvault_vault[0].resource_id : var.key_vault.id
```
Comment (Nepomuceno) [inline main.tf]: if the resource itself uses resource_id it is better for you to also use resource_id and not to id so if somoene created the resource outside they could just pass in the whole resource and it should work.

File: variables.tf [hunk 4]
```diff
@@ -226,12 +226,34 @@ variable "tags" {
   description = "(Optional) Tags of the resource."
 }
 
-variable "key_vault_id" {
+variable "shared_subnet_id" {
   type        = string
```
Comment (Nepomuceno) [inline variables.tf:192]: YOu should use the same approach here that you used for keyvault and storage account. to pass the resurce in.

**Review by Nepomuceno — COMMENTED** [review]
Great job just some small comments. :)

File: examples/privatelink/main.tf [hunk 5]
```diff
@@ -52,10 +52,19 @@ locals {
   firewall_subnet_address_space         = local.core_services_vnet_subnets[1]
   bastion_subnet_address_prefix         = local.core_services_vnet_subnets[2]
   shared_services_subnet_address_prefix = local.core_services_vnet_subnets[3]
-  dns_zones = toset([
+  azureml_dns_zones = toset([
     "privatelink.api.azureml.ms",
     "privatelink.notebooks.azure.net",
   ])
```
Comment (Nepomuceno) [inline examples/privatelink/main.tf:53]: Are those values always these values if so we would not need someone to pass them in would we ?

File: examples/privatelink/main.tf [hunk 6]
```diff
@@ -52,10 +52,19 @@ locals {
...
+  key_vault_endpoints = toset([
+    "vaultcore",
+  ])
+  storage_account_endpoints = toset([
+    "blob",
+    "file",
+    "queue",
+    "table",
+  ])
```
Comment (Nepomuceno) [inline examples/privatelink/main.tf:65]: Same wuestion with the ones from azureml if they are fixed better to just see if oneone says private or not no ?

File: examples/privatelink/main.tf [hunk 7]
```diff
@@ -115,5 +156,18 @@ module "azureml" {
...
+    endpoints = local.key_vault_endpoints
+    private_dns_zone_resource_ids = [azurerm_private_dns_zone.key_vault_dns_zones["vaultcore"].id]
+  }
+
+  storage_account = {
+    endpoints = local.storage_account_endpoints
+    private_dns_zone_resource_ids = [azurerm_private_dns_zone.storage_dns_zones["blob"].id, 
+                                      azurerm_private_dns_zone.storage_dns_zones["file"].id, 
+                                      azurerm_private_dns_zone.storage_dns_zones["queue"].id, 
+                                      azurerm_private_dns_zone.storage_dns_zones["table"].id]
```
Comment (Nepomuceno) [inline examples/privatelink/main.tf]: Here you are creating a dependency of someone creating an endpoint for all the possible resources of azure storage imagine they did not want to use azure queue or something like that. ( not sure it is possible ) but if the are required to use all them it is not an option to pass the array in.

**Review by Nepomuceno — COMMENTED** [review]
Just some changes on the zones not sure they are actually needed.

**Review by Nepomuceno — APPROVED** [review]
LGTM

**Pull request merged**